ouroboros-network/src/Ouroboros/Network/Protocol/TxSubmission/Client.hs

```diff
@@ -0,0 +1,106 @@
+{-# LANGUAGE DataKinds           #-}
+{-# LANGUAGE GADTs               #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE PolyKinds           #-}
+module Ouroboros.Network.Protocol.TxSubmission.Client where
+
+import Protocol.Core
+import Ouroboros.Network.Protocol.TxSubmission.Type
+
+import Data.List.NonEmpty (NonEmpty (..))
+import Pipes (Producer)
+import qualified Pipes
+
+newtype TxSubmissionClient (n :: Nat) tx err m a = TxSubmissionClient {
+    runTxSubmissionClient :: m (TxSubmission n tx err m a)
+  }
+
+instance Functor m => Functor (TxSubmissionClient n tx err m) where
+  fmap f (TxSubmissionClient cli) = TxSubmissionClient $ (fmap . fmap) f cli
+
+data TxSubmission (n :: Nat) tx err m a where
+  TxSubmission
+    :: LessEqualThan (Succ Zero) n ~ True
+    => tx
+    -> TxSubmissionClient (Pred n) tx err m a
+    -> TxSubmission n tx err m a
+  TxSubmissionDone
+    :: a
+    -> TxSubmissionErrHandler err m a
+    -> TxSubmission n tx err m a
```

Comment:
Is this the same as

```
data TxSubmission (n :: Nat) tx err m a where
  TxSubmission
    :: tx
    -> TxSubmissionClient n tx err m a
    -> TxSubmission (S n) tx err m a
  TxSubmissionDone
    :: a
    -> TxSubmissionErrHandler err m a
    -> TxSubmission n tx err m a
```

?

The use of a `LessEqualThan` constraint I think is quite more awkward.

Reply:
No, it's not; the original will not allow to construct a `TxSubmission` that submits more than `n` transactions.

Reply:
A `TxSubmission n tx err m a` is either `TxSubmissionDone`, which submits 0, or `TxSubmission`, which submits one and then at most `n-1` more. It's a bit like the `Fin` type

```
data Fin (n : Nat) where
  FZ : Fin (S k)
  FS : Fin k -> Fin (S k)
```

Reply:
Yes, indeed it is very like `Fin (n :: Nat)`.

Reply:
Taking advantage of the duality between `(->) x` and `(,) x`, and using our naming convention:
```
data Fin (n :: Nat) (f :: * -> * -> *) x where
  FZ :: Fin (Succ k) f x
  FS :: f x (Fin k f x) -> Fin (Succ k) f x

direct
  :: Fin n (->) x
  -> Fin n (,)  x
  -> ()
direct FZ FZ = ()
direct (FS f) (FS (x, fs)) = direct (f x) fs
```